pyquil/noise.py

```diff
@@ -0,0 +1,551 @@
+##############################################################################
+# Copyright 2018 Rigetti Computing
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+"""
+Module for creating and verifying noisy gate and readout definitions.
+"""
+import warnings
+from collections import namedtuple
+
+import numpy as np
+
+from pyquil.parameters import format_parameter
+from pyquil.quilbase import Pragma, Gate, Qubit as QuilQubit
+
+
+INFTY = float("inf")
+"Used for infinite coherence times."
+
+
+_KrausModel = namedtuple("_KrausModel", ["gate", "params", "targets", "kraus_ops", "fidelity"])
+
+
+class KrausModel(_KrausModel):
+    """
+    Encapsulate a single gate's noise model.
+
+    :ivar str gate: The name of the gate.
+    :ivar Sequence[float] params: Optional parameters for the gate.
+    :ivar Sequence[int] targets: The target qubit ids.
+    :ivar Sequence[np.array] kraus_ops: The Kraus operators (must be square complex numpy arrays).
+    :ivar float fidelity: The average gate fidelity associated with the Kraus map relative to the
+        ideal operation.
+    """
+
+    @staticmethod
+    def unpack_kraus_matrix(m):
+        """
+        Helper to optionally unpack a JSON compatible representation of a complex Kraus matrix.
+
+        :param Union[list,np.array] m: The representation of a Kraus operator. Either a complex
+            square matrix (as numpy array or nested lists) or a pair of real matrices (as numpy
```

Comment:
This matrix pair business is to support JSON serialization?

Reply:
Yeah, I slightly changed the wording to add that info